File: src/algorithms/binary-trees/node-depths/solution1.py
# O(n) time | O(h) depth
# n = # nodes
# h = tree height
def nodeDepths(root, depth = 0):
	if root is None:
		return 0
	return depth + nodeDepths(root.left, depth + 1) + nodeDepths(root.right, depth + 1)
    

# This is the class of the input binary tree.
class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

File: src/algorithms/binary-trees/node-depths/solution1.ts
// O(n) time | O(h) space
// n = # nodes
// h = tree height
export function nodeDepths(root: BinaryTree) {
  let sumOfDepths = 0
  const stack: { node: BinaryTree | null; depth: number }[] = [
    { node: root, depth: 0 }
  ]
  while (stack.length > 0) {
    const { node, depth } = stack.pop()!
    if (node === null) continue
    sumOfDepths += depth
    stack.push({ node: node.left, depth: depth + 1 })
    stack.push({ node: node.right, depth: depth + 1 })
  }
  return sumOfDepths
}

// This is the class of the input binary tree.
class BinaryTree {
  value: number
  left: BinaryTree | null
  right: BinaryTree | null

  constructor(value: number) {
    this.value = value
    this.left = null
    this.right = null
  }
}
